### PR TITLE
MAINT: Vendor npymath as git submodule and vendor npyrandom directly

### LINKED ISSUE #17498: Remove use of `libnpymath` and `libnpyrandom`
We should get this done for `1.11.0`. Using static libraries from `numpy` is unhealthy. Let me summarize the main problems and link some related issues.

From https://github.com/scipy/scipy/pull/17481#issuecomment-1328263356:
_The main issue is using a static library at all. This is bad, mainly for these two reasons:_
1. _It is not guaranteed that `numpy` and `scipy` are built with the same compilers. And that leads to potential issues that are hard to predict, as we've experienced multiple times already. Especially on Windows, where we build scipy with mingw and numpy with MSVC. That was the reason why using the `vc142` instead of the `vc141` toolchain in numpy 1.21.5-1.22.1 led to complex issues and that change had to be reverted._
2. _Cross-compiling tends to break. We haven't officially supported that in the past, but even in the one case where we do it (macOS x64-64 to arm64) it goes pretty horribly wrong: we stuff the x86-64 `libnpymath.a` into the macOS scipy wheel. This silently worked, because the required symbols were already loaded by numpy itself. Now that they're being removed from `numpy`, we're all of a sudden loading them from `libnpymath.a` and they have the wrong architecture._

For more details, see https://github.com/numpy/numpy/issues/20880.

Here are some related issues where the use of `libnpmath` showed up as a problem:

- gh-13323 (contains `libnpymath.a(npy_math.o), could not parse object file`)
- Cross-compiling issue: gh-16783
- Another cross-compiling issue: https://github.com/scipy/scipy/issues/14812#issuecomment-1204871264
- Regression because scipy 1.8/1.9 wheels contain x86-64 instructions in macOS arm64 wheel: https://github.com/numpy/numpy/issues/22673

What we should do is described in https://github.com/numpy/numpy/issues/20880#issuecomment-1327171950. Basically, vendor the sources for the functions/typedefs we still need, and rebuild them as part of the SciPy build. This is probably best done by refactoring the numpy sources a bit so there's a standalone C99 compat library with only the complex typedefs, complex functions from C99 that MSVC doesn't provide, and the blocklisted functions. And then add a command-line switch to the numpy meson build to only build that library (use a meson [disabler object](https://mesonbuild.com/Disabler.html) for that). And include numpy as a [meson subproject](https://mesonbuild.com/Subprojects.html#using-a-subproject) in the scipy source tree - this is kinda like a git submodule - so we can build it as needed, and don't actually need to vendor anything.

I don't have the bandwidth to push this forward now, but should be able to do so in Q1'23.

### PR DESCRIPTION
**Pull request by lysnikolaou**
Closes gh-17498

### CODE REVIEW

**Comment by lucascolley** [comment]
x-ref gh-17498

**Comment by rgommers** [comment]
Those source files aren't shipped by NumPy yet, so that cannot work I think - as also indicated by CI failures:
```
 scipy/meson.build:80:30: ERROR: File ../test/lib/python3.9/site-packages/numpy/core/include/../../random/src/distributions/distributions.c does not exist.
```

We are supporting ~4 numpy versions back, so rather than trying to ship the C sources as part of NumPy, they should be vendored somehow (either directly or via a submodule). I was suggesting above that what we need is so minimal (one struct, one function) that it may be possible to vendor the relevant bits directly.

File: scipy/special/meson.build [hunk 1]
```diff
@@ -331,11 +331,7 @@ py3.extension_module('_ufuncs',
   ],
   c_args: [cython_c_args, Wno_maybe_uninitialized],
   include_directories: ['../_lib', '../_build_utils/src'],
-  dependencies: [
-    lapack,
-    npymath_lib,
```
Comment (rgommers) [inline scipy/special/meson.build:336]: This must still be needed right? You were replacing it with the new subproject, but that seems to have gone missing from this PR now.

**Comment by matthew-brett** [comment]
> I got successfully (locally) Windows 32-bit build with [lysnikolaou@4829d69](https://github.com/lysnikolaou/scipy/commit/4829d69cc3c143f62f5d6751c2f895e15c39fa3d) out of the box!

Ah - excellent - but I guess with the same ndimage test failures as before?

**Comment by carlkl** [comment]
The same failures unfortunately. Not related to this PR however.

**Comment by lysnikolaou** [comment]
This is now updated to use numpy/libnpymath, instead of my test repo. A round of review would be nice, so that we can merge it.

**Comment by lucascolley** [comment]
The `cp311 (build sdist + wheel), full, no pythran` fail is unrelated.

File: meson.build [hunk 2]
```diff
@@ -142,4 +142,10 @@ if use_pythran
   xsimd_dep = dependency('xsimd', required: false)
 endif
 
+libnpymath_dep = dependency(
```
Comment (rgommers) [inline meson.build:145]: Maybe this deserves a comment like:
```
# We don't expect `libnpymath` to be packaged separately, but looking for a standalone
# install is useful in case of local development (don't have to touch the git submodule)
```

At least I assume that that's the intent. I haven't checked the `libnpymath` repo yet. Are there pkg-config files for it, or is that planned? If not, the above doesn't work and hence we don't need to try looking for the dependency.

File: scipy/stats/_rcont/_rcont.c [hunk 3]
```diff
@@ -10,11 +10,11 @@
 
   Author: Hans Dembinski
 */
-
+#include <math.h>
```
Comment (rgommers) [inline scipy/stats/_rcont/_rcont.c:13]: I thought at first that this was the wrong include order - (C standard library headers before `Python.h`). However, `Python.h` is not included anymore. It'd be useful to note in a comment (or `libnpyrandom/README.md`?) that none of that code should be depending on NumPy or Python.

**Review by rgommers — COMMENTED** [review]
This looks quite good! Only one nontrivial comment so far. I'll have a closer look at the included C code, but should be safe since it's a sync from numpy code.

**Comment by rgommers** [comment]
If you push another commit, could you include `[wheel build]` in the first line of your commit message? That way we run all wheel builds too, and ensure that's all green.

File: meson.build (hunk 2 above)
Comment (lysnikolaou) [reply meson.build:145]: In my mind this was to ensure that this continues to work even if `libnpymath` is shipped another way. Right now, it does not ship any `pkg-config` files, but it could in the future. Either way I'll remove this and go with the `subproject` call for now and maybe change it in the future in case we do something else.

Comment (rgommers) [reply meson.build:145]: That sounds good, thanks. It's easy enough to change back later once someone actually needs this.

File: scipy/stats/_rcont/_rcont.c (hunk 3 above)
Comment (lysnikolaou) [reply scipy/stats/_rcont/_rcont.c:13]: That's true. I'll add a note to `libnpyrandom/README.md`.

**Comment by tylerjereddy** [comment]
@lysnikolaou can you summarize for me the relationship (if any) between this change and NumPy 2.0.0? As I recall, Ralf wanted this merged a few release ago, but then things got tricky with various discussion upstream.

I'm thinking of branching for `1.13.0` (NumPy 2.0.0 support release) ~Sunday the 17th.

**Comment by lysnikolaou** [comment]
> @lysnikolaou can you summarize for me the relationship (if any) between this change and NumPy 2.0.0? As I recall, Ralf wanted this merged a few release ago, but then things got tricky with various discussion upstream.
> 
> I'm thinking of branching for `1.13.0` (NumPy 2.0.0 support release) ~Sunday the 17th.

*TL;DR This is not needed for 2.0 compatibility.*

The bigger picture is that we want to eventually stop shipping static libraries with NumPy. We're working with the assumption that SciPy is one of the few, if not the only, downstream user of the static libraries, hence this PR that adds support for the new way of doing things.

However, this change did not make it into 2.0 because of challenges with complex numbers. I'm actively working on a solution for these problems, and hope that everything will be ready for 2.1. In case that goes well, `npymath` will be moved out of the NumPy source tree and it will be maintained in a [separate repo](https://github.com/numpy/libnpymath). NumPy will still ship with the `npymath` static library for backwards-compatibility reasons, although we would like to eventually sunset that, and the best way to go forward for downstream would be to vendor `npymath` as a git submodule instead of relying on NumPy for it.

For `npyrandom`, the purpose of this PR is to remove any dependency on it, and instead vendor the few needed files directly. This way, we can start figuring out a plan on how to stop shipping the `npyrandom` static library as well.

**Comment by rgommers** [comment]
The `libnpyrandom` part of this PR was split off and merged in gh-20413. The rest accumulated quite a few conflicts. With all the code rewrites into C++ going on in `scipy.special`, some of the complex type related issues there may have already been solved.

For a next step here, I'd recommend another PR splitting off whatever is mergeable next. I think it the removal of `libnpymath` does not have to be all at once, right? If so, then just adding the new subproject with the changes to `sparsetools` and `ndimage` would work. Although - that code doesn't use `libnpymath` but only `npy_math.h` - so is that actually a useful/necessary change?

**Comment by ev-br** [comment]
A gentle ping, as this seems to have slipped on everyone's radar. Would be great to move this along.
